File: patterns/fractal_breakout.py
# trade-signals-web/patterns/fractal_breakout.py

import pandas as pd
from .base_pattern import BasePattern

# Глобальный кэш для фрактальных сигналов
_fractal_signal_cache = {}

def reset_fractal_cache():
    """Сброс кэша фрактальных сигналов перед каждым сканом"""
    global _fractal_signal_cache
    _fractal_signal_cache = {}

class FractalBreakoutPattern(BasePattern):
    """Пробой фрактальной линии"""
    
    def __init__(self):
        super().__init__("Пробой фрактальной линии (19)", confidence="medium")
    
    def get_min_candles(self) -> int:
        return 40
    
    def find_fractals(self, df, left=9, right=9):
        """Поиск фракталов"""
        highs = df['high'].values
        lows = df['low'].values
        upper_fractals = []
        lower_fractals = []
        
        for i in range(left, len(df) - right):
            # Верхний фрактал
            is_upper = True
            for j in range(1, left + 1):
                if i - j >= 0 and highs[i] <= highs[i - j]:
                    is_upper = False
                    break
            for j in range(1, right + 1):
                if i + j < len(highs) and highs[i] <= highs[i + j]:
                    is_upper = False
                    break
            if is_upper:
                upper_fractals.append((i, highs[i]))
            
            # Нижний фрактал
            is_lower = True
            for j in range(1, left + 1):
                if i - j >= 0 and lows[i] >= lows[i - j]:
                    is_lower = False
                    break
            for j in range(1, right + 1):
                if i + j < len(lows) and lows[i] >= lows[i + j]:
                    is_lower = False
                    break
            if is_lower:
                lower_fractals.append((i, lows[i]))
        
        return upper_fractals, lower_fractals
    
    def detect(self, df: pd.DataFrame) -> dict | None:
        global _fractal_signal_cache
        
        upper_fractals, lower_fractals = self.find_fractals(df, left=9, right=9)
        last_idx = len(df) - 1
        last_close = df['close'].iloc[-1]
        
        # Сигнал на покупку (BUY) - пробой верхней линии вверх
        if len(upper_fractals) >= 2:
            p1_idx, p1_price = upper_fractals[-2]
            p2_idx, p2_price = upper_fractals[-1]
            
            delta_idx = p2_idx - p1_idx
            if delta_idx != 0:
                slope = (p2_price - p1_price) / delta_idx
                
                # Игнорируем горизонтальные линии
                if abs(slope) < 1e-6:
                    return None
                
                prev_line = p1_price + slope * (last_idx - 1 - p1_idx)
                curr_line = p1_price + slope * (last_idx - p1_idx)
                
                pair_key = f"BUY_{p1_idx}_{p2_idx}"
                
                if pair_key not in _fractal_signal_cache:
                    if len(df) > 1:
                        prev_close = df['close'].iloc[-2]
                        if prev_close <= prev_line and last_close > curr_line:
                            _fractal_signal_cache[pair_key] = True
                            return {
                                "signal": "BUY",
                                "description": "BUY при пробое трендовой линии по фракталам (19)",
                                "confidence": self.confidence
                            }
        
        # Сигнал на продажу (SELL) - пробой нижней линии вниз
        if len(lower_fractals) >= 2:
            p1_idx, p1_price = lower_fractals[-2]
            p2_idx, p2_price = lower_fractals[-1]
            
            delta_idx = p2_idx - p1_idx
            if delta_idx != 0:
                slope = (p2_price - p1_price) / delta_idx
                
                if abs(slope) < 1e-6:
                    return None
                
                prev_line = p1_price + slope * (last_idx - 1 - p1_idx)
                curr_line = p1_price + slope * (last_idx - p1_idx)
                
                pair_key = f"SELL_{p1_idx}_{p2_idx}"
                
                if pair_key not in _fractal_signal_cache:
                    if len(df) > 1:
                        prev_close = df['close'].iloc[-2]
                        if prev_close >= prev_line and last_close < curr_line:
                            _fractal_signal_cache[pair_key] = True
                            return {
                                "signal": "SELL",
                                "description": "SELL при пробое трендовой линии по фракталам (19)",
                                "confidence": self.confidence
                            }
        
        return None

File: js/patterns/fractal_breakout.js
// js/patterns/fractal_breakout.js

// Кэш с временем жизни
const signalCache = new Map();
const CACHE_TTL = {
    '1h': 4 * 60 * 60 * 1000,     // 4 часа
    '1d': 3 * 24 * 60 * 60 * 1000, // 3 дня
    '1wk': 14 * 24 * 60 * 60 * 1000, // 2 недели
    '1mo': 30 * 24 * 60 * 60 * 1000  // 30 дней
};

function cleanExpiredCache() {
    const now = Date.now();
    for (const [key, data] of signalCache.entries()) {
        if (now - data.timestamp > data.ttl) {
            signalCache.delete(key);
        }
    }
}

export function clearFractalCache() {
    signalCache.clear();
}

export class FractalBreakoutPattern {
    name = 'Пробой фрактальной линии (19)';
    confidence = 'medium';

    findFractals(candles, left = 9, right = 9) {
        const highs = candles.map(c => c.high);
        const lows = candles.map(c => c.low);
        const upperFractals = [];
        const lowerFractals = [];

        for (let i = left; i < candles.length - right; i++) {
            let isUpper = true;
            for (let j = 1; j <= left; j++) {
                if (highs[i] <= highs[i - j]) { isUpper = false; break; }
            }
            for (let j = 1; j <= right; j++) {
                if (highs[i] <= highs[i + j]) { isUpper = false; break; }
            }
            if (isUpper) upperFractals.push({ index: i, price: highs[i] });

            let isLower = true;
            for (let j = 1; j <= left; j++) {
                if (lows[i] >= lows[i - j]) { isLower = false; break; }
            }
            for (let j = 1; j <= right; j++) {
                if (lows[i] >= lows[i + j]) { isLower = false; break; }
            }
            if (isLower) lowerFractals.push({ index: i, price: lows[i] });
        }

        return { upperFractals, lowerFractals };
    }

    detect(candles, timeframe = '1d') {
        if (candles.length < 40) return null;

        cleanExpiredCache();

        const { upperFractals, lowerFractals } = this.findFractals(candles);
        const lastIdx = candles.length - 1;
        const lastClose = candles[lastIdx].close;

        const ttl = CACHE_TTL[timeframe] || CACHE_TTL['1d'];

        // BUY сигнал
        if (upperFractals.length >= 2) {
            const p1 = upperFractals[upperFractals.length - 2];
            const p2 = upperFractals[upperFractals.length - 1];

            const deltaIdx = p2.index - p1.index;
            if (deltaIdx !== 0) {
                const slope = (p2.price - p1.price) / deltaIdx;
                if (Math.abs(slope) < 1e-6) return null;

                const prevLine = p1.price + slope * (lastIdx - 1 - p1.index);
                const currLine = p1.price + slope * (lastIdx - p1.index);

                const key = `BUY_${p1.index}_${p2.index}`;
                
                const cached = signalCache.get(key);
                const now = Date.now();
                
                if (!cached || (now - cached.timestamp > cached.ttl)) {
                    const prevClose = candles[lastIdx - 1].close;
                    if (prevClose <= prevLine && lastClose > currLine) {
                        signalCache.set(key, {
                            timestamp: now,
                            ttl: ttl,
                            signal: 'BUY'
                        });
                        return {
                            signal: 'BUY',
                            description: 'BUY при пробое трендовой линии по фракталам (19)',
                            confidence: this.confidence
                        };
                    }
                }
            }
        }

        // SELL сигнал
        if (lowerFractals.length >= 2) {
            const p1 = lowerFractals[lowerFractals.length - 2];
            const p2 = lowerFractals[lowerFractals.length - 1];

            const deltaIdx = p2.index - p1.index;
            if (deltaIdx !== 0) {
                const slope = (p2.price - p1.price) / deltaIdx;
                if (Math.abs(slope) < 1e-6) return null;

                const prevLine = p1.price + slope * (lastIdx - 1 - p1.index);
                const currLine = p1.price + slope * (lastIdx - p1.index);

                const key = `SELL_${p1.index}_${p2.index}`;
                
                const cached = signalCache.get(key);
                const now = Date.now();
                
                if (!cached || (now - cached.timestamp > cached.ttl)) {
                    const prevClose = candles[lastIdx - 1].close;
                    if (prevClose >= prevLine && lastClose < currLine) {
                        signalCache.set(key, {
                            timestamp: now,
                            ttl: ttl,
                            signal: 'SELL'
                        });
                        return {
                            signal: 'SELL',
                            description: 'SELL при пробое трендовой линии по фракталам (19)',
                            confidence: this.confidence
                        };
                    }
                }
            }
        }

        return null;
    }
}
